aten/src/ATen/BatchingRegistrations.cpp

```diff
@@ -0,0 +1,34 @@
+#include <ATen/BatchedTensorImpl.h>
+#include <ATen/BatchingRules.h>
+#include <torch/library.h>
+
+namespace at {
+
+/*
+ * Operator Registrations for BatchedTensor.
+ * Contains some glue to hook up the batching rules to BatchedTensorImpl.
+ */
+
+Tensor BatchedTensor_sum(const Tensor& self, IntArrayRef dim, bool keepdim, c10::optional<ScalarType> dtype) {
+  Tensor self_, result_;
+  BatchDimsRef self_bdims;
+  BatchDims result_bdims;
+
+  std::tie(self_, self_bdims) = unpackBatched(self);
+  std::tie(result_, result_bdims) = sum_batching_rule(self_, self_bdims, dim, keepdim, dtype);
```

Comment:
How come `sum_batching_rule` doesn't just return a BatchedTensor from the getgo?

Reply:
This is probably what you are referring to with 

> Batching rules and utility functions act on "unpacked BatchedTensors".
> This is to avoid tensor allocation overhead (one alternative is having
> these act on BatchedTensors). Unfortunately this leads to more verbose
> and confusing code. I've considered adding a special
> "BatchedTensorData" struct that only holds a Tensor and BatchDims and
> has helper methods defined on it but I'm not sure if that is better.

But I don't understand how passing things in expanded form helps avoid allocations.

Reply:
Is the allocation avoidance specifically for the case when we mixed batched and non-batched tensors?

Reply:
I wanted the batching rules to look similar to their TorchScript variants. Right now I am envisioning that we have these batching rules in C++, and then a separate TorchScript set. The TorchScript batching rules should take in tensors and return tensors, because it would attempt to rewrite the graph using regular tensors.

However, that is not set in stone and I probably won't worry about it for ~2 months. I'm not sure if it is better to make the simplifying assumption that "batching rules should return BatchedTensors" and then worry about how TorchScript batching rules will look like in the future

Reply:
Yeah, the allocation avoidance advice is specifically for the case when we mix batched and non-batched -- wrapping the non-batched tensor in a BatchedTensorImpl would be unnecessary. But maybe that is premature optimization.

The main reason for the batching rules to look like this is because I wanted them to look similar to however we would add them in TorchScript in the future, but maybe this is also premature optimization

Reply:
OK, definitely worth writing down in code what you are planning for in the future, no matter what.

The form of the TorchScript translations is an interesting problem. I actually think you should consider working out how to make TorchScript vmap rules accept batch tensors, and not the deconstructed tensors. I imagine how I would go about doing this in a good old fashioned compiler and the easiest way to eliminate vmap from a program is to simply inline all the implementations of vmapped operations into the use sites. To do this inlining, you have construct a semantic correspondence between the batched tensors (in the source program) with the batched tensors of the inputs to the batching rules (the function being inlined). 

So no matter what, you need a way to tell that the pair of Tensor and BDims corresponded to a batch tensor in the source program. So you could have some rule that constructs this correspondence... or you could have just passed them in as batched tensors to begin with!

What I am not so sure about is how convenient it will be to do this sort of thing in TorchScript as the compiler is setup today.

Reply:
> So no matter what, you need a way to tell that the pair of Tensor and BDims corresponded to a batch tensor in the source program. So you could have some rule that constructs this correspondence... or you could have just passed them in as batched tensors to begin with!

That's a good argument for passing the tensors in as BatchedTensors. I am not sure how convenient it is to do this in TorchScript either, but it is something I'll need to figure out down the line